src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+                new QCallTypeHandle(ref groupType),
+                &NewExternalTypeEntry,
+                &NewProxyTypeEntry,
+                &context);
+
+            _externalTypeMap = context.ExternalTypeMap;
+            _proxyTypeMap = context.ProxyTypeMap;
+        }
+
+        public IReadOnlyDictionary<string, Type> GetExternalTypeMap() => _externalTypeMap;
+        public IReadOnlyDictionary<Type, Type> GetProxyTypeMap() => _proxyTypeMap;
+
+        private abstract class LazyTypeLoadDictionary<TKey> : IReadOnlyDictionary<TKey, Type> where TKey : notnull
+        {
+            protected abstract bool TryGetOrLoadType(TKey key, [NotNullWhen(true)] out Type? type);
+
+            public Type this[TKey key]
+            {
+                get
+                {
+                    if (!TryGetOrLoadType(key, out Type? type))
+                    {
+                        ThrowHelper.ThrowKeyNotFoundException(key);
+                    }
+
+                    return type;
+                }
+            }
+
+            public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out Type value) => TryGetOrLoadType(key, out value);
+
+            // Not supported to avoid exposing TypeMap entries in a manner that
+            // would violate invariants the Trimmer is attempting to enforce.
+            public IEnumerable<TKey> Keys => throw new NotSupportedException();
+            public IEnumerable<Type> Values => throw new NotSupportedException();
+            public int Count => throw new NotSupportedException();
+            public bool ContainsKey(TKey key) => throw new NotSupportedException();
+            public IEnumerator<KeyValuePair<TKey, Type>> GetEnumerator() => throw new NotSupportedException();
+            IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
+        }
+
+        [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+        private sealed class DelayedType
+        {
+            private Type? _type;
+            public DelayedType(string assemblyName, string typeName)
+            {
+                AssemblyName = assemblyName;
+                TypeName = typeName;
+                _type = null;
+            }
+
+            public string AssemblyName { get; }
+            public string TypeName { get; }
+
+            public Type GetOrLoadType()
+            {
+                if (_type is null)
+                {
+                    lock (this)
+                    {
+                        Assembly targetAssembly = Assembly.Load(AssemblyName);
+                        _type = targetAssembly.GetType(TypeName, throwOnError: true)!;
+                    }
+                }
+                return _type;
+            }
+
+            public bool RepresentsSameType(DelayedType other)
+            {
+                return string.Equals(AssemblyName, other.AssemblyName, StringComparison.Ordinal)
```

Comment:
The actual assembly may end up being the same, even through the assembly name is different (e.g. due to different version or due to forwarders).

Is it a problem? Would it be better to avoid comparing the assembly name here so that we have fewer corner cases?

Reply:
I can probably do that.